usr/man/_vax.d/man2/chown.2
.\" SCCSID: @(#)chown.2	8.1	9/11/90
.TH chown 2
.SH Name
chown, fchown \- change owner and group of a file
.SH Syntax
.nf
#include <sys/types.h>
.PP
chown(\fIpath, owner, group\fP)
char *\fIpath\fP;
uid_t \fIowner\fP;
gid_t \fIgroup\fP;
.PP
fchown(\fIfd, owner, group\fP)
int \fIfd\fP;
uid_t \fIowner\fP;
gid_t \fIgroup\fP;
.fi
.SH Description
.NXR "chown system call"
.NXR "fchown system call"
.NXR "file" "changing owner"
.NXR "file" "changing group"
The 
.PN chown
and
.PN fchown
system calls change the owner and group of the file named by 
.I path
or referenced by
.I fd .
Only the superuser can change the owner
of a file.  Other users can change the group-id of a file that they own
to another group to which they belong.
.PP
If you specify \-1 in \fIowner\fP or \fIgroup,\fP the corresponding owner-id or
group-id of
the file is unchanged.
.PP
The
.PN chown
system call clears the set-user-id and set-group-id bits on
the file when it returns successfully, unless the call is made by the
superuser. Clearing these bits when a file's owner is changed protects the
file from remaining set-user-id or set-group-id after being modified.
If a file, specifically a program, remained set-user-id
or set-group-id after being modified, that file could allow
unauthorized access to other files or accounts.
.PP
You should use the
.PN fchown
system call 
with the file locking primitives because 
.PN fchown
preserves any locks you previously obtained with the
.PN flock
system call.  For more information about file locking, see the
.MS flock 2 
reference page.
.SH Return Values
The
.PN chown
and 
.PN fchown
calls return zero if the operation is successful; if an error occurs they return
\-1 and store a more specific
error code in the global variable \fIerrno\fP.
.SH Environment
.SS System Five 
Differs from the System V definition in that only
the superuser can change the ownership of a file.  In addition, 
ELOOP is a possible error condition.
.SS POSIX
When your program is compiled in the POSIX environment, the
.I owner
argument is of type 
.I uid_t, 
and the
.I group
argument is of type 
.I gid_t.
.SH Diagnostics
The
.PN chown
system call fails and the file is unchanged under the following
conditions:
.TP 15
[ENOTDIR]
A component of the path prefix is not a directory.
.TP 15
[ENAMETOOLONG]
A component of a pathname exceeded 255 characters, or an
entire pathname exceeded 1023 characters.
.TP 15
[ENOENT]
The named file does not exist.
.TP 15
[EACCES]
Search permission is denied for a component of the path prefix.
.TP 15
[EPERM]
The effective user-id is not the superuser.
.TP 15
[EROFS]
The named file resides on a read-only file system.
.TP 15
[EFAULT]
The pathname
points outside the process's allocated address space.
.TP 15
[ELOOP]
Too many symbolic links are encountered in translating the pathname.
.TP 15
[EIO]
An I/O error occurs while reading from or writing to the file system.
.TP 15
[ESTALE]
The 
.I fd
argument is invalid because the file referred
to by that file handle no longer exists or has been revoked.
.PP
The
.PN fchown
system call fails if:
.TP 15
[EBADF]
The
.I fd
argument does not refer to a valid descriptor.
.TP 15
[EINVAL]
The
.I fd
argument refers to a socket, not a file.
.TP 15
[EPERM]
The effective user-id is not the superuser.
.TP 15
[EROFS]
The named file resides on a read-only file system.
.TP 15
[EIO]
An I/O error occurred while reading from or writing to the
file system.
.TP 15
[ETIMEDOUT]
A connect request or remote file operation fails because the connected party
does not properly respond after a period of time that is dependent on
the communications protocol.
.SH See Also
chmod(2), flock(2)


usr/etc/chown.d/chown.c
#ifndef lint
static	char	*sccsid = "@(#)chown.c	4.1	Ultrix		7/2/90";
#endif lint
/************************************************************************
 *									*
 *			Copyright (c) 1989 by				*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   This software is  derived  from  software  received  from  the	*
 *   University    of   California,   Berkeley,   and   from   Bell	*
 *   Laboratories.  Use, duplication, or disclosure is  subject  to	*
 *   restrictions  under  license  agreements  with  University  of	*
 *   California and with AT&T.						*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/
/************************************************************************
 *			Modification History				*
 * 002 Richard Hart, 9-Apr-87						*
 *     Added an extra argument to Perror to amke it more informative.	*
 * 001 Richard Hart, 9-Apr-87						*
 *     Copied from 4.3 BSD:						*
 *		chown.c      5.6 (Berkeley) 5/29/86			*
 ************************************************************************/

/*
 * chown [-fR] uid[.gid] file ...
 */

#include <stdio.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <pwd.h>
#include <sys/dir.h>
#include <grp.h>
#include <strings.h>

struct	passwd *pwd;
struct	passwd *getpwnam();
struct	stat stbuf;
int	uid;
int	status;
int	fflag;
int	rflag;

main(argc, argv)
	char *argv[];
{
	register int c, gid;
	register char *cp, *group;
	struct group *grp;

	argc--, argv++;
	while (argc > 0 && argv[0][0] == '-') {
		for (cp = &argv[0][1]; *cp; cp++) switch (*cp) {

		case 'f':
			fflag++;
			break;

		case 'R':
			rflag++;
			break;

		default:
			fatal(255, "unknown option: %c", *cp);
		}
		argv++, argc--;
	}
	if (argc < 2) {
		fprintf(stderr, "usage: chown [-fR] owner[.group] file ...\n");
		exit(-1);
	}
	gid = -1;
	group = index(argv[0], '.');
	if (group != NULL) {
		*group++ = '\0';
		if (!isnumber(group)) {
			if ((grp = getgrnam(group)) == NULL)
				fatal(255, "unknown group: %s",group);
			gid = grp -> gr_gid;
			(void) endgrent();
		} else if (*group != '\0')
			gid = atoi(group);
	}
	if (!isnumber(argv[0])) {
		if ((pwd = getpwnam(argv[0])) == NULL)
			fatal(255, "unknown user id: %s",argv[0]);
		uid = pwd->pw_uid;
	} else
		uid = atoi(argv[0]);
	for (c = 1; c < argc; c++) {
		/* do stat for directory arguments */
		if (lstat(argv[c], &stbuf) < 0) {
			status += Perror("access", argv[c]);
			continue;
		}
		if (rflag && ((stbuf.st_mode&S_IFMT) == S_IFDIR)) {
			status += chownr(argv[c], uid, gid);
			continue;
		}
		if (chown(argv[c], uid, gid)) {
			status += Perror("change ownership of", argv[c]);
			continue;
		}
	}
	exit(status);
}

isnumber(s)
	char *s;
{
	register c;

	while(c = *s++)
		if (!isdigit(c))
			return (0);
	return (1);
}

chownr(dir, uid, gid)
	char *dir;
{
	register DIR *dirp;
	register struct direct *dp;
	struct stat st;
	char savedir[1024];
	int ecode;
	extern char *getwd();

	if (getwd(savedir) == (char *)0)
		fatal(255, "%s", savedir);
	/*
	 * Change what we are given before doing it's contents.
	 */
	if (chown(dir, uid, gid) < 0 && Perror("change ownership of", dir))
		return (1);
	if (chdir(dir) < 0) {
		Perror("chdir to", dir);
		return (1);
	}
	if ((dirp = opendir(".")) == NULL) {
		Perror("open", dir);
		return (1);
	}
	dp = readdir(dirp);
	dp = readdir(dirp); /* read "." and ".." */
	ecode = 0;
	for (dp = readdir(dirp); dp != NULL; dp = readdir(dirp)) {
		if (lstat(dp->d_name, &st) < 0) {
			ecode = Perror("access", dp->d_name);
			if (ecode)
				break;
			continue;
		}
		if ((st.st_mode&S_IFMT) == S_IFDIR) {
			ecode = chownr(dp->d_name, uid, gid);
			if (ecode)
				break;
			continue;
		}
		if (chown(dp->d_name, uid, gid) < 0 &&
		    (ecode = Perror("change ownership of", dp->d_name)))
			break;
	}
	closedir(dirp);
	if (chdir(savedir) < 0)
		fatal(255, "can't change back to %s", savedir);
	return (ecode);
}

error(fmt, a)
	char *fmt, *a;
{

	if (!fflag) {
		fprintf(stderr, "chown: ");
		fprintf(stderr, fmt, a);
		putc('\n', stderr);
	}
	return (!fflag);
}

fatal(status, fmt, a)
	int status;
	char *fmt, *a;
{

	fflag = 0;
	(void) error(fmt, a);
	exit(status);
}

Perror(operation, s)
	char *operation, *s;
{

	if (!fflag) {
		fprintf(stderr, "chown: can't %s ", operation);
		perror(s);
	}
	return (!fflag);
}
